pkg/util/fswatch/fswatch.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2022 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fswatch provies Watch(), a utility function to watch a filesystem path.
+package fswatch
+
+import (
+	"time"
+
+	"agones.dev/agones/pkg/util/runtime"
+	"gopkg.in/fsnotify.v1"
+)
+
+var logger = runtime.NewLoggerWithSource("fswatch")
+
+// Watch watched the filesystem path `path`. When anything changes, changes are
+// batched for the period `batchFor`, then `processEvent` is called.
+func Watch(path string, batchFor time.Duration, processEvent func()) (func(), error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	cancelChan := make(chan struct{})
+	cancel := func() {
+		close(cancelChan)
+		_ = watcher.Close()
+	}
+	if err := watcher.Add(path); err != nil {
+		cancel()
+		return nil, err
+	}
+
+	go batchWatch(batchFor, watcher.Events, watcher.Errors, cancelChan, processEvent, func(error) {
+		logger.WithError(err).Errorf("error watching %s", path)
+	})
+	return cancel, nil
+}
+
+// batchWatch: watch for events; when an event occurs, keep draining events for duration `batchFor`, then call processEvent().
+// Intended for batching of rapid-fire events where we want to process the batch once, like filesystem update notifications.
+func batchWatch(batchFor time.Duration, events chan fsnotify.Event, errors chan error, cancelChan chan struct{}, processEvent func(), onError func(error)) {
+	// Pattern shamelessly stolen from https://blog.gopheracademy.com/advent-2013/day-24-channel-buffering-patterns/
```

Comment:
ooooh! Fancy!

Reply:
I almost started to hand roll it, and then I had this tickle that I've seen this pattern before... :)